File: .NET/Survey/DisabledControlTooltipDispTest/WindowsFormsApp1/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Reflection.Emit;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace WindowsFormsApp1
{
    /// <summary>
    /// コントロールが非活性でもツールチップを表示するテスト
    /// 
    /// <参考>
    /// https://stackoverflow.com/questions/1732140/displaying-tooltip-over-a-disabled-control
    /// </summary>
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();

            var td = new ToolTipOnDisabledControl();
            //td.SetToolTip(this.button2, "2" );
            td.SetToolTip(this.button2, this.toolTip1, "2");    //test
        }

        private void checkBox1_CheckedChanged(object sender, EventArgs e)
        {
            button1.Enabled = !button1.Enabled;

            if (!button1.Enabled)
            {
                var td = new ToolTipOnDisabledControl();
                //td.SetToolTip(this.button1, "1");
                td.SetToolTip(this.button1, this.toolTip1, "1");    //test
            }
            else
            {
                this.toolTip1.SetToolTip(this.button1, "Hello button1");//test
            }

        }

        private void Form1_Load(object sender, EventArgs e)
        {
            //this.button1.Location = new System.Drawing.Point(143, 102);
            //this.button1.Size = new System.Drawing.Size(270, 91);

            //var rect = new Rectangle();
            //rect.X = button1.Location.X;
            //rect.Y = button1.Location.Y;
            //rect.Width = button1.Size.Width;
            //rect.Height = button1.Size.Height;

            //label2.Text = rect.ToString();

            //label12.Text = checkBox1.Bounds.ToString();
        }

        private void Form1_MouseMove(object sender, MouseEventArgs e)
        {
            return;

            label6.Text = "";

            label4.Text = e.Location.ToString();    // クライアント座標

            //// マウスの現在の位置を取得
            //Point mousePosition = Cursor.Position;      // スクリーン座標
            //label8.Text = mousePosition.ToString();

            //// フォーム上の座標に変換
            //Point formPosition = this.PointToClient(mousePosition); // クライアント座標
            //label10.Text = formPosition.ToString();

            //// 座標にある子コントロールを取得
            //Control control = this.GetChildAtPoint(formPosition); // GetChildAtPointは完全一致したときでないと取得できない。

            foreach (Control childControl in this.Controls)
            {
                if (childControl.Enabled == false)
                {
                    if ((e.Location.X >= childControl.Location.X &&
                         e.Location.X <= (childControl.Location.X + childControl.Size.Width))
                          &&
                        (e.Location.Y >= childControl.Location.Y &&
                         e.Location.Y <= (childControl.Location.Y + childControl.Size.Height)))
                    {
                        label6.Text = childControl.Name;
                        label6.Text = this.toolTip1.GetToolTip(childControl);
                        this.toolTip1.ShowAlways = true;
                        this.toolTip1.Show(label6.Text, childControl, childControl.Width / 2, childControl.Height / 2);
                        break;
                    }
                    //else
                    //{
                    //    label6.Text = "";
                    //    this.toolTip1.ShowAlways = false;
                    //    this.toolTip1.Hide(childControl);
                    //    break;
                    //}
                    //break;
                }
            }
        }

    }

    /// <summary>
    /// // Reference example
    ///  var td = new ToolTipOnDisabledControl();
    ///  this.checkEdit3.Enabled = false;
    ///  td.SetTooltip(this.checkEdit3, "tooltip for disabled");
    /// </summary>
    public class ToolTipOnDisabledControl
    {
        #region Fields and Properties

        private Control enabledParentControl;

        private bool isShown;

        public Control TargetControl { get; private set; }

        public string TooltipText { get; private set; }
        public ToolTip ToolTip { get; }
        #endregion

        #region Public Methods
        public ToolTipOnDisabledControl()
        {
            this.ToolTip = new ToolTip();
        }

        //public void SetToolTip(Control targetControl, string tooltipText = null)
        public void SetToolTip(Control targetControl, ToolTip tip,  string tooltipText = null)//test
        {
            this.TargetControl = targetControl;
            if (string.IsNullOrEmpty(tooltipText))
            {
                this.TooltipText = this.ToolTip.GetToolTip(targetControl);
            }
            else
            {
                this.TooltipText = tooltipText;

                tip.SetToolTip(targetControl, "");  // test
            }

            if (targetControl.Enabled)
            {
                this.enabledParentControl = null;
                this.isShown = false;
                this.ToolTip.SetToolTip(this.TargetControl, this.TooltipText);
                return;
            }

            this.enabledParentControl = targetControl.Parent;
            while (!this.enabledParentControl.Enabled && this.enabledParentControl.Parent != null)
            {
                this.enabledParentControl = this.enabledParentControl.Parent;
            }

            if (!this.enabledParentControl.Enabled)
            {
                throw new Exception("Failed to set tool tip because failed to find an enabled parent control.");
            }

            this.enabledParentControl.MouseMove += this.EnabledParentControl_MouseMove;
            this.TargetControl.EnabledChanged += this.TargetControl_EnabledChanged;
        }

        public void Reset()
        {
            if (this.TargetControl != null)
            {
                this.ToolTip.Hide(this.TargetControl);
                this.TargetControl.EnabledChanged -= this.TargetControl_EnabledChanged;
                this.TargetControl = null;
            }

            if (this.enabledParentControl != null)
            {
                this.enabledParentControl.MouseMove -= this.EnabledParentControl_MouseMove;
                this.enabledParentControl = null;
            }

            this.isShown = false;
        }
        #endregion

        #region Private Methods
        private void EnabledParentControl_MouseMove(object sender, MouseEventArgs e)
        {
            if (e.Location.X >= this.TargetControl.Left &&
                e.Location.X <= this.TargetControl.Right &&
                e.Location.Y >= this.TargetControl.Top &&
                e.Location.Y <= this.TargetControl.Bottom)
            {
                if (!this.isShown)
                {
                    this.ToolTip.Show(this.TooltipText, this.TargetControl, this.TargetControl.Width / 2, this.TargetControl.Height / 2, this.ToolTip.AutoPopDelay);
                    this.isShown = true;
                }
            }
            //else if(this.isShown)
            else
            {
                this.ToolTip.Hide(this.TargetControl);
                this.isShown = false;
            }
        }

        private void TargetControl_EnabledChanged(object sender, EventArgs e)
        {
            if (TargetControl.Enabled)
            {
                TargetControl.EnabledChanged -= TargetControl_EnabledChanged;
                enabledParentControl.MouseMove -= EnabledParentControl_MouseMove;
            }
        }
        #endregion
    }
}

File: .NET/Survey/DisabledControlTooltipDispTest/VB/WindowsApp1/Form1.vb
Public Class Form1
	Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load
		Dim td As New ToolTipOnDisabledControl()
		td.SetToolTip(Me.button2, Me.ToolTip1, "This tooltip is Disabled for Button2")

	End Sub



	Private Sub checkBox1_CheckedChanged(sender As Object, e As EventArgs) Handles checkBox1.CheckedChanged
		button1.Enabled = Not button1.Enabled

		If Not button1.Enabled Then
			Dim td As New ToolTipOnDisabledControl()
			'td.SetToolTip(this.button1, "1");
			'test
			td.SetToolTip(Me.button1, Me.ToolTip1, "This tooltip is Disabled for Button1")
		Else
			'test
			Me.ToolTip1.SetToolTip(Me.button1, "This tooltip is Enabled for Button1")
		End If
	End Sub

	''' <summary>
	''' // Reference example
	'''  var td = new ToolTipOnDisabledControl();
	'''  this.checkEdit3.Enabled = false;
	'''  td.SetTooltip(this.checkEdit3, "tooltip for disabled");
	''' </summary>
	Public Class ToolTipOnDisabledControl
#Region "Fields and Properties"

		Private enabledParentControl As Control

		Private isShown As Boolean

		Public Property TargetControl() As Control

		Public Property TooltipText() As String
		Public ReadOnly Property ToolTip() As ToolTip
#End Region

#Region "Public Methods"
		Public Sub New()
			Me.ToolTip = New ToolTip()

			Dim timer As New Timer()
			timer.Interval = 100 ' 100ミリ秒ごとに更新
			AddHandler timer.Tick, AddressOf Timer_Tick
			timer.Start()
		End Sub

		Private Sub Timer_Tick(ByVal sender As Object, ByVal e As EventArgs)
			' マウスカーソルの座標を取得して表示する
			'Dim cursorPosition As Point = Cursor.Position
			'Me.Label1.Text = $"X: {cursorPosition.X}, Y: {cursorPosition.Y}"

			' マウスカーソルのスクリーン座標を取得
			Dim screenPoint As Point = Cursor.Position

			' コントロールに対して座標変換を行う
			Dim clientPoint As Point = Me.enabledParentControl.PointToClient(screenPoint)

			If clientPoint.X >= Me.TargetControl.Left AndAlso
			   clientPoint.X <= Me.TargetControl.Right AndAlso
			   clientPoint.Y >= Me.TargetControl.Top AndAlso
			   clientPoint.Y <= Me.TargetControl.Bottom Then
				If Not Me.isShown Then
					Me.ToolTip.Show(
						Me.TooltipText,
						Me.TargetControl,
						Me.TargetControl.Width / 2, Me.TargetControl.Height / 2,
						Me.ToolTip.AutoPopDelay)
					Me.isShown = True
				End If
			Else
				'else if(this.isShown)
				Me.ToolTip.Hide(Me.TargetControl)
				Me.isShown = False
			End If
		End Sub

		'public void SetToolTip(Control targetControl, string tooltipText = null)
		Public Sub SetToolTip(targetControl As Control, tip As ToolTip, Optional tooltipText As String = Nothing)
			'test
			Me.TargetControl = targetControl
			If String.IsNullOrEmpty(tooltipText) Then
				Me.TooltipText = Me.ToolTip.GetToolTip(targetControl)
			Else
				Me.TooltipText = tooltipText

				' test
				tip.SetToolTip(targetControl, "")
			End If

			If targetControl.Enabled Then
				Me.enabledParentControl = Nothing
				Me.isShown = False
				Me.ToolTip.SetToolTip(Me.TargetControl, Me.TooltipText)
				Return
			End If

			Me.enabledParentControl = targetControl.Parent
			While Not Me.enabledParentControl.Enabled AndAlso Me.enabledParentControl.Parent IsNot Nothing
				Me.enabledParentControl = Me.enabledParentControl.Parent
			End While

			If Not Me.enabledParentControl.Enabled Then
				Throw New Exception("Failed to set tool tip because failed to find an enabled parent control.")
			End If

			AddHandler Me.enabledParentControl.MouseMove, AddressOf Me.EnabledParentControl_MouseMove
			AddHandler Me.TargetControl.EnabledChanged, AddressOf Me.TargetControl_EnabledChanged
		End Sub

		Public Sub Reset()
			If Me.TargetControl IsNot Nothing Then
				Me.ToolTip.Hide(Me.TargetControl)
				RemoveHandler Me.TargetControl.EnabledChanged, AddressOf Me.TargetControl_EnabledChanged
				Me.TargetControl = Nothing
			End If

			If Me.enabledParentControl IsNot Nothing Then
				RemoveHandler Me.enabledParentControl.MouseMove, AddressOf Me.EnabledParentControl_MouseMove
				Me.enabledParentControl = Nothing
			End If

			Me.isShown = False
		End Sub
#End Region

#Region "Private Methods"
		Private Sub EnabledParentControl_MouseMove(sender As Object, e As MouseEventArgs)
			'If e.Location.X >= Me.TargetControl.Left AndAlso
			'   e.Location.X <= Me.TargetControl.Right AndAlso
			'   e.Location.Y >= Me.TargetControl.Top AndAlso
			'   e.Location.Y <= Me.TargetControl.Bottom Then
			'	If Not Me.isShown Then
			'		Me.ToolTip.Show(
			'			Me.TooltipText,
			'			Me.TargetControl,
			'			Me.TargetControl.Width / 2, Me.TargetControl.Height / 2,
			'			Me.ToolTip.AutoPopDelay)
			'		Me.isShown = True
			'	End If
			'Else
			'	'else if(this.isShown)
			'	Me.ToolTip.Hide(Me.TargetControl)
			'	Me.isShown = False
			'End If
		End Sub

		Private Sub TargetControl_EnabledChanged(sender As Object, e As EventArgs)
			If TargetControl.Enabled Then
				RemoveHandler TargetControl.EnabledChanged, AddressOf TargetControl_EnabledChanged
				RemoveHandler enabledParentControl.MouseMove, AddressOf EnabledParentControl_MouseMove
			End If
		End Sub
#End Region
	End Class
End Class
